lib/diskoverflow/sorted.go

```diff
@@ -0,0 +1,274 @@
+// Copyright (C) 2018 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package diskoverflow
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"sort"
+)
+
+const suffixLength = 8
+
+// Sorted is a disk-spilling container, that sorts added values by the
+// accompanying keys.
+// You must not add new values after calling PopFirst/-Last.
+type Sorted interface {
```

Comment:
This is a sorted map. I guess we could just live with one implementation of map, which can be sorted by default? Do we really need the second one?

Reply:
You can't get items by key and sorting is expensive and pointless for maps.

Reply:
Not really, leveldb already does it. I am not sure why we're not just using inmemory level db, and when that is full spill to on disk level db, then your map is sorted, and has pop first pop last.